polkadot/node/core/backing/src/lib.rs

```diff
@@ -981,6 +990,74 @@ async fn handle_active_leaves_update<Context>(
 	Ok(())
 }
 
+macro_rules! try_runtime_api {
+	($x: expr) => {
+		match $x {
+			Ok(x) => x,
+			Err(err) => {
+				// Only bubble up fatal errors.
+				error::log_error(Err(Into::<runtime::Error>::into(err).into()))?;
+
+				// We can't do candidate validation work if we don't have the
+				// requisite runtime API data. But these errors should not take
+				// down the node.
+				return Ok(None)
+			},
+		}
+	};
+}
+
+#[overseer::contextbounds(ApprovalVoting, prefix = self::overseer)]
+async fn core_index_from_statement<Sender>(
+	sender: &mut Sender,
+	relay_parent: Hash,
+	statement: &SignedFullStatementWithPVD,
+) -> Result<Option<CoreIndex>, Error>
+where
+	Sender: SubsystemSender<RuntimeApiMessage> + Clone,
+{
+	let parent = relay_parent;
+
+	let (validator_groups, group_rotation_info) = request_validator_groups(parent, sender)
+		.await
+		.await
+		.map_err(Error::RuntimeApiUnavailable)?
+		.map_err(Error::FetchValidatorGroups)?;
+
+	let cores = request_availability_cores(parent, sender)
+		.await
+		.await
+		.map_err(Error::RuntimeApiUnavailable)?
+		.map_err(Error::FetchAvailabilityCores)?;
+
+	let compact_statement = statement.as_unchecked();
+	let candidate_hash = CandidateHash(*compact_statement.unchecked_payload().candidate_hash());
+
+	gum::trace!(target: LOG_TARGET, ?group_rotation_info, ?statement, ?validator_groups, ?cores, ?candidate_hash, "Extracting core index from statement");
+
+	Ok(core_index_from_statement_inner(&cores, &validator_groups, &group_rotation_info, statement))
+}
+
+pub(crate) fn core_index_from_statement_inner(
+	cores: &[CoreState],
+	validator_groups: &[Vec<ValidatorIndex>],
+	group_rotation_info: &GroupRotationInfo,
+	statement: &SignedFullStatementWithPVD,
+) -> Option<CoreIndex> {
+	let statement_validator_index = statement.validator_index();
+	for (group_index, group) in validator_groups.iter().enumerate() {
+		for validator_index in group {
+			if *validator_index == statement_validator_index {
+				return Some(
+					group_rotation_info.core_for_group(GroupIndex(group_index as u32), cores.len()),
```

Comment:
We should also check that the paraid of the core matches the candidate in the statement.

Reply:
With the [upcoming exposure of claimqueue](https://github.com/orgs/paritytech/projects/119/views/20?pane=issue&itemId=42349591) this will need to be changed to check for any occurrence in the claim queue of the core. @tdimitrov

Reply:
done